uwblueprint/planet-read
# Store and retrieve files

## Notion ticket link
[Store and retrieve files](https://www.notion.so/uwblueprintexecs/a658c933d18c48d4b43fd99bda2bd3b4?v=4a47edc475b34bb99b25ff1fae2a5507&p=9352edb12bb8460294b9b505d36ee523)


## Implementation description
* Added Files model (id, path)
* Added POST /files endpoint to allow users to upload a single file
* Added GET /files/<id> to allow users to retrieve path to file
* Files are uploaded to a folder in the container and not to the cloud at the moment


## Steps to test
1. Add an `UPLOAD_PATH=/var/lib/file_uploads` environment variable to point to the directory that files will be uploaded to 
2. Tear down containers and rebuild:
```
$ docker-compose down --volumes
$ docker-compose up --build
```
3. Go into `/backend/python/app/models/__init__.py` and change the `erase_db_and_sync = False` to `True`
4. Verify the Files table is created:
```
$ docker exec -it <db-container-id> psql -U postgres -d planet-read
$ \dt
```
![image](https://user-images.githubusercontent.com/32009013/119275030-d07f3c80-bbe0-11eb-8151-6b1fe6747a75.png)

5. Login and obtain the access_token for a user
```
curl --location --request POST 'localhost:5000/auth/login' \
--header 'Content-Type: application/json' \
--data-raw '{
    "email": <USER EMAIL ADDRESS>,
    "password": "<USER PASSWORD>
}'
```

6. Try to upload a file to the database (use the access_token from above):
```
curl --location --request POST 'http://localhost:5000/files' \
--header 'Authorization: Bearer  <ACCESS_TOKEN>' \
--form 'file=@"<PATH_TO_FILE>"'
```
Observe that the response looks something like:
```
{
    "id": 1,
    "path": "./app/files/my_file.pdf"
}
```

7. Try to retrieve the file path from the database:
```
curl --location --request GET 'http://localhost:5000/files/1' \
--header 'Limit: 1' \
--header 'Authorization: Bearer <ACCESS_TOKEN>' \
--header 'Content-Type: application/json' 
```
Observe that the response looks something like:
```
{
    "id": 1,
    "path": "./app/files/my_file.pdf"
}
```
8. Verify that the files are stored correctly in the volume:
```
$ docker run -it --rm -v planet-read_file_uploads:/vol busybox ls -l /vol
```

## What should reviewers focus on?
* Verifying that the endpoints are structured correctly
* Actual file changes (ran the linter so the git diff looks really messy sorry about that!):
  * `backend/python/app/models/file.py`
  * `backend/python/app/resources/file_dto.py`
  * `backend/python/app/rest/file_routes.py`
  * `backend/python/app/services/implementations/file_service.py` 
  * `backend/python/app/services/interfaces/file_service.py`

## Checklist
